Fix: docker exec hangs indefinitely when reading from stdin

bcxpro opened the pull request:
This PR fixes the improper handling of half-closed connection behavior when attaching stdin to a container. The proper approach requires:
- The docker client receives a half-close of the connection from client to daemon (stdin)
- The connection from daemon to client (stdout, stderr) remains open until no more data is to be received

The first part of the issue is resolved by enabling MessageMode on Windows named pipes. This ensures that when stdin ends, an EOF is received on the reader, allowing proper half-closing of the connection.

The previous implementation used stdlib's httputil.DockerProxy to proxy from Windows named pipe to WSL hvsock. This proxy does not use CloseWrite on hijacked connections, preventing proper handling of this use case. By using Close() instead of CloseWrite(), both connection directions are closed simultaneously, causing the client to miss pending stdout/stderr content after stdin EOF.

This commit introduces a simpler custom `util.ReverseProxy` to replace httputil.DockerProxy, implementing proper handling of half-closed connections with explicit support for:
- HTTP protocol upgrades
- Half-close TCP connection management
- Precise stream handling for Docker API interactions

Closes #2094

Nino-K commented:
@bcxpro thanks for your contributions, it looks like there are some [issues](https://github.com/rancher-sandbox/rancher-desktop/actions/runs/12246608513/job/34209894006?pr=7896) with the "Test" step in the CI. Once they are addressed I will gladly take a look at your PR. Thanks

bcxpro commented:
> @bcxpro, thanks for your contribution. It looks like there are some [issues](https://github.com/rancher-sandbox/rancher-desktop/actions/runs/12246608513/job/34209894006?pr=7896) with the "Test" step in the CI. Once they are addressed, I will gladly look at your PR.
> 
> Also, once you can run `yarn install && yarn lint` before pushing your PR, ensure you can address the lint issues. Thanks

@Nino-K I had missed the changes to a file in the commit. I just updated the PR with it. It should build now.

Regarding the linter, I fixed all the issues except one. Honestly, I cannot figure out exactly what response body is complaining about in that line.

bcxpro commented:
I just have seen that this one may be indirectly fixing #3239

matejkramny commented:
Hi, looks like we worked on the same problem simultaneously and solved it differently.

I'm curious why change the windows side? Attaching to containers from Windows didn't have this problem

this is my understanding from looking at how rancher runs:
![mermaid-diagram-2024-12-11-161227](https://github.com/user-attachments/assets/556efd16-f973-438a-a948-49f35e589d8e)

The issue seemed to be only at the linux proxy (red box in the diagram)

bcxpro commented:
> Hi, looks like we worked on the same problem simultaneously and solved it differently.
> 
> I'm curious why change the windows side? Attaching to containers from Windows didn't have this problem


Hi @matejkramny,

Your diagram is correct, at least as far as I understand it.

Initially, a couple of months ago, I tried to run _kind_ on Rancher Desktop and it failed. Looking into it, I discovered that the problem was the _kind_ tool passing configuration on Windows via stdin, and the docker commands would get stuck.
I found issue #2094, which exactly matched the problem I had on Windows. The issue can be reproduced with:

```
echo ok | docker run -i --rm alpine cat
```

When you run this, "ok" is echoed back to the terminal, but the command gets stuck instead of exiting and returning the correct exit code.

I started troubleshooting that part. The first problem was that the Windows named pipe was not signaling the EOF of the data received via stdin. As a result, the Docker CLI was stuck, always listening for more data. The data was echoed back, but that was it. Initially, I thought I had found the solution with this line in server_windows.go:

```
pipeConfig := &winio.PipeConfig{MessageMode: true}
```

Everything seemed to work—the "ok" was echoed back, and the program was no longer stuck.

However, _kind_ still did not work. So I continued troubleshooting and found that the _http.ReverseProxy_ used in `wsl-helper serve` to proxy from the Windows named pipe to the Hyper-V vsock did not properly half-close the connections. The Docker API for attach requires that the stdin end signals the end of stdin by half-closing the connection. But _http.ReverseProxy_ just closes the connection in both directions.
This causes issues with commands like:


```
cat largefile.txt | docker run -i --rm alpine cat
```

The problem then is that the command's output is truncated. The command ends but it does not echo back everything.

The reason is that for the short period of time that the connection to the hvsock remains open traffic flows in all directions, and as soon as it is closed, the stdout/stderr data cannot flow to the client anymore. That is the reason why at first just with that like in server_windows.go the command `echo ok | docker run -i --rm alpine cat` seemed to work.

Then the problem was the proxy. I initially built a very simple L4-level proxy that just piped both connections properly. But the current solution needed to leverage the capabilities of hooks in _http.ReverseProxy_ to modify requests and responses (mungers). A transparent proxy wouldn't suffice, so an L7 proxy was needed that could make proper use of CloseWrite to half-close connections.

In this PR, I wrote a minimalistic HTTP reverse proxy that can handle both half-closes and the "mungers".

With that, the solution to my problem was finished, and I submitted this PR a couple of days ago.

Yesterday, I saw #3239 related to WSL. Testing it with this PR applied, it looks like the issue is solved. This is most likely because the `wsl-helper serve` used to pipe from the /var/run/docker.sock in WSL to the /mnt/wsl/...sock uses the same code. The replacement of the proxy that I did to solve the problem in Windows also solves the WSL issue (specifically, the half-close problem).

matejkramny commented:
@bcxpro that's great, thank you for explaining!

> Then the problem was the proxy. I initially built a very simple L4-level proxy that just piped both connections properly. But the current solution needed to leverage the capabilities of hooks in http.ReverseProxy to modify requests and responses (mungers). A transparent proxy wouldn't suffice, so an L7 proxy was needed that could make proper use of CloseWrite to half-close connections.

I think our solutions diverge here, at L4 I read ahead and match the request to `/attach` and conditionally continue piping at L4 or to the munger-capable L7 server.

For interest, go maintainers [don't seem to have plans to fix the ReverseProxy](https://github.com/golang/go/issues/70754)

jandubois commented:
> For interest, go maintainers [don't seem to have plans to fix the ReverseProxy](https://github.com/golang/go/issues/70754)

Not sure this is true; that issue was just closed as a duplicate of https://github.com/golang/go/issues/35892, which is still open, but doesn't seem to have any recent action on it either.

matejkramny commented:
It wasn't explicitly mentioned but it's an open bug from 2019






---- On Wed, 11 Dec 2024 18:39:32 +0000 ***@***.*** wrote ----



For interest, go maintainers don't seem to have plans to fix the ReverseProxy

Not sure this is true; that issue was just closed as a duplicate of golang/go#35892, which is still open, but doesn't seem to have any recent action on it either.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.

MrWako commented:
Hi @jandubois, @matejkramny, @bcxpro - I've tried to pick up https://github.com/golang/go/issues/35892 again. I couldn't see any particular objections to addressing the issue so hoping that with a bit of perseverance we can get a fix in and avoid having to maintain a custom proxy within the rancher desktop code.

jandubois commented:
> I've tried to pick up [golang/go#35892](https://github.com/golang/go/issues/35892) again. I couldn't see any particular objections to addressing the issue so hoping that with a bit of perseverance we can get a fix in

@MrWako Thanks for picking this up!

> and avoid having to maintain a custom proxy within the rancher desktop code.

I think it is fine to use a custom proxy in Rancher Desktop until this is fixed upstream because we don't know how long it would take, or if upstream will accept it at all.

Nino-K reviewed (commented):
I haven't looked at the `reverse_proxy.go`, I will do that shortly.

bcxpro commented on src/go/wsl-helper/pkg/dockerproxy/start.go:
```diff
@@ -177,7 +177,21 @@ func handleConnection(conn net.Conn, dockerPath string) {
+		panic("daemon connection does not implement HalfReadCloseWriter")
+	} else {
+		xDockerConn = x
+	}
```
I it with your code.

bcxpro commented on src/go/wsl-helper/pkg/dockerproxy/util/pipe.go:
```diff
@@ -17,11 +17,11 @@ limitations under the License.
 
-// Pipe bidirectionally between two streams.
-func Pipe(c1, c2 io.ReadWriteCloser) error {
+func Pipe(c1, c2 HalfReadWriteCloser) error {
```
There are two Pipe functions, one that is ons `networking/go/utils` and the one that I modified in `dockerproxy/util/pipe.go`. This one is not used anywhere else.

bcxpro commented on src/go/wsl-helper/pkg/dockerproxy/serve.go:
```diff
@@ -124,7 +120,6 @@ func Serve(endpoint string, dialer func() (net.Conn, error)) error {
 			}
 			return nil
 		},
-		ErrorLog: log.New(logWriter, "", 0),
```
In my first implementation of the alternative proxy I did not implement logging. I just updated the code with logging and restored that line.

Nino-K commented on src/go/wsl-helper/pkg/dockerproxy/util/reverse_proxy.go:
```diff
@@ -0,0 +1,277 @@
+
+	// Write the response status code and headers and flush it immediately
+	w.WriteHeader(backendResponse.StatusCode)
+	flusher, ok := w.(http.Flusher)
```
You've checked if the `ResponseWriter` is a `http.Flusher` twice—once inside the `periodicHttpFlush` and again right before the `WriteHeader` call. If possible, it would be cleaner to do this check once and handle the flushing logic in one place. At the minimum, I think checking the flusher inside the goroutine might not be needed.

bcxpro replied:
I removed the checking form the goroutine because as I stated in my previous comment, at the time the goroutine is launched now the check has been performed succesfully.

Nino-K commented on src/go/wsl-helper/pkg/dockerproxy/util/reverse_proxy.go:
```diff
@@ -0,0 +1,277 @@
+}
+
+func (p *ReverseProxy) logf(format string, args ...any) {
+	logger := p.ErrorLog
```
It might be unnecessary to declare this variable. You could just do:
```
	if p.ErrorLog == nil {
		p.ErrorLog = log.Default()
	}

	p.ErrorLog.Printf(format, args...)
```

bcxpro replied:
I’ve thought about it, and I would prefer to keep the current implementation for two main reasons:

Avoiding Side Effects: The `ErrorLog` field is public, and modifying it inside `logf` introduces a side effect that changes the state of a public field the first time that the function is invoked. By using a local variable, the current implementation avoids this side effect and keeps the function more predictable.

Consistency with the standard library: This code is intended to replace the `http.ReverseProxy` struct from the standard Go library, and I’ve tried to maintain consistency with its behavior. The standard library’s ReverseProxy does not modify the `ErrorLog` field when logging, and I’d like to preserve that behavior to make it easier to swap back to the standard implementation if the half closing bug is eventually solved.

Nino-K commented on src/go/wsl-helper/pkg/dockerproxy/util/reverse_proxy.go:
```diff
@@ -0,0 +1,277 @@
+	_, err = io.Copy(w, backendResponse.Body)
+	if err != nil {
+		proxy.logf("failed to stream the response body to the client: %v", err.Error())
+		return
```
I'm surprised Linter didn't get mad over this return 😄  because your function would have returned regardless. Also, this is totally optional but you could simplify this to something like: 
```
	// Stream the response body back to the client
	if _, err = io.Copy(w, backendResponse.Body); err != nil {
		proxy.logf("failed to stream the response body to the client: %v", err.Error())
	}
```

Nino-K commented on src/go/wsl-helper/pkg/dockerproxy/util/reverse_proxy.go:
```diff
@@ -0,0 +1,277 @@
+	// Cast writer to safely hijack the connection
+	hijacker, ok := w.(http.Hijacker)
+	if !ok {
+		const errMsg = "client response writer does not support http.Hijacker"
```
I would suggest introducing the following helper function, which can significantly simplify your code, something like the below:

```
// sendError is a helper method for sending consistent error messages to the client
func (proxy *ReverseProxy) sendError(w http.ResponseWriter, msg string, statusCode int) {
	proxy.logf(msg)
	http.Error(w, msg, statusCode)
}
```
Then you can just call it like:
```
proxy.sendError(w, "client response writer does not support http.Hijacker", http.StatusInternalServerError)
```
And that goes for all other 7 instances below.

bcxpro replied:
I implemented the `sendError` function as you proposed it. In some places in which I had used that `errMsg` pattern I had removed sending the error because it was not needed (basically I was trying to send HTTP response messages once the response message had already been sent)

For the rest of the cases, I used `sendError`

Nino-K commented on src/go/wsl-helper/pkg/dockerproxy/util/reverse_proxy.go:
```diff
@@ -0,0 +1,277 @@
+	// - bufferedClientConn: A buffered reader/writer for any pending data
+	clientConn, bufferedClientConn, err := hijacker.Hijack()
+	if err != nil {
+		proxy.logf("cannot hijack client connection: %v", err.Error())
```
For these ones you can just:
```
proxy.sendError(w, "cannot hijack client connection: "+err.Error(), http.StatusInternalServerError)
```
or
```
proxy.sendError(w, fmt.Errorf("cannot hijack client connection: %v", err).Error(), http.StatusInternalServerError)
```
Either one should work.

bcxpro replied:
Done

Nino-K commented on src/go/wsl-helper/pkg/dockerproxy/util/reverse_proxy.go:
```diff
@@ -0,0 +1,277 @@
+	}
+
+	// Cast backend and client connections to HalfReadWriteCloser
+	var xBackendConn HalfReadWriteCloser
```
Also, this suggestion is totally optional but you can simply your type assertion by introducing a helper function like below:
```
// assertHalfReadWriteCloser ensures the connection implements HalfReadWriteCloser
func assertHalfReadWriteCloser(conn net.Conn) (HalfReadWriteCloser, error) {
	halfRWConn, ok := conn.(HalfReadWriteCloser)
	if !ok {
		return nil, fmt.Errorf("connection does not implement HalfReadWriteCloser")
	}
	return halfRWConn, nil
}
```
And then: 
```
	xBackendConn, err := assertHalfReadWriteCloser(backendConn)
	if err != nil {
		proxy.sendError(w, err.Error(), http.StatusInternalServerError)
		return
	}
	xClientConn, err := assertHalfReadWriteCloser(clientConn)
	if err != nil {
		proxy.sendError(w, err.Error(), http.StatusInternalServerError)
		return
	}
```
Whether or not you introduce `assertHalfReadWriteCloser`, it would be helpful to consider renaming the variable `x` to something more descriptive for clarity.

bcxpro replied:
This is the part in which I removed sending the HTTP error because at that time it is too late, the connection has been hijacked and the response status code has already been sent. So this code is now shorter.

I have not implemented the assert function because as the code is now I think it adds a bit of complexity IMHO. It saves two lines of code and adds a 7-line function that is used only there.

bcxpro replied:
I removed the checking form the goroutine because as I stated in my previous comment, at the time the goroutine is launched now the check has been performed succesfully.

bcxpro replied:
I also renamed the x.. variables.. The original idea was that the `xSomething` variable was the casted version of `something`. I changed the name and I hope it works; naming is hard.

bcxpro replied:
I have also renamed the x.. variables.. The original idea was that the `xSomething` variable was the casted version of `something`. I changed the name and I hope it works; naming is hard.

Nino-K replied:
> I have also renamed the x.. variables.. The original idea was that the `xSomething` variable was the casted version of `something`. I changed the name and I hope it works; naming is hard.

I think you may have forgotten to push your changes, I still see the `x` name in there.

bcxpro replied:
Sorry, this is a confusion on my part. I though you were referring to `xBackendConn` and `xClientConn`.

I’m still getting familiar with Go’s naming practices, so I thought `x` was fine for such a short-lived variable. I’ve now changed it to `halfCloser` to make it more descriptive.

Nino-K reviewed (changes requested):
Thanks for making the changes. I’ve added some feedback, mostly to help reduce redundancy. Overall, the functionality looks reasonable.

bcxpro commented on src/go/wsl-helper/pkg/dockerproxy/util/reverse_proxy.go:
```diff
@@ -0,0 +1,277 @@
+	// Establish a connection to the backend using a custom Dial method
+	backendConn, err := proxy.Dial("", "")
+	if err != nil {
+		proxy.logf("failed to connect to the backend: %v", err.Error())
```
Fixed

Nino-K commented on src/go/wsl-helper/pkg/dockerproxy/util/reverse_proxy.go:
```diff
@@ -0,0 +1,260 @@
+	// Critical for Docker API's stream-based communication model
+	if err := Pipe(halfCloserClientConn, halfCloserBackendConn); err != nil {
+		proxy.logf("piping client to backend failed: %v", err)
+		return
```
Just FYI, you can remove this return and it's unnecessary.

Nino-K commented on src/go/wsl-helper/pkg/dockerproxy/util/reverse_proxy.go:
```diff
@@ -0,0 +1,260 @@
+	// Cast backend and client connections to HalfReadWriteCloser
+	var halfCloserBackendConn HalfReadWriteCloser
+	var halfCloserClientConn HalfReadWriteCloser
+	if x, ok := backendConn.(HalfReadWriteCloser); !ok {
```
The `x` is still here :(

bcxpro replied:
Done

Nino-K reviewed (changes requested):
Thanks for making all the changes it looks good, I just had two minor comments we can go ahead with a merge after those.

bcxpro commented:
@Nino-K I just found a bug in the code caused by a race condition that happens when Flush and Write are invoked at the same time. I have already worked on a solution that I am now testing and I will be submitting it most probably later today.

bcxpro commented:
As previously reported today, I identified a race condition between Flush and Write operations. Due to the periodic nature of the Flush operation, it would eventually coincide with a Write execution.

This issue caused container log following to fail after some time. The problem can be reproduced with these commands (it requires some patience to observe the failure):

```sh
docker run --rm --name mylogger -d chentex/random-logger:latest 10 5
docker logs mylogger -f
```

To resolve this, I've implemented a new flushedWriter struct that wraps an io.Writer and protects it with a mutex, ensuring that Flush and Write operations cannot execute simultaneously. I've also relocated the flushing goroutine code into this new structure.

Testing confirms that this implementation resolves the race condition.

Nino-K reviewed (commented):
Thanks for making the changes and addressing the race condition. I had a chance to do a final review, and everything seems to be okay. I also had a chance to test this. I just left you a single comment regarding string concatenation. Once you address this, we can go ahead and merge the PR. Thanks again!

bcxpro commented on src/go/wsl-helper/pkg/dockerproxy/util/reverse_proxy.go:
```diff
@@ -0,0 +1,301 @@
+	defer backendConn.Close()
+
+	// Create a new HTTP request with the same headers
+	url := targetProtocol + hostHeaderValue + r.RequestURI
```
Because those are two constants the Go compliler optimizes it. You can see it here in Compiler Explorer: https://godbolt.org/z/c5Gvd9ovP

The assignment in this sample function is optimized.
```go
const (
	hostHeaderValue = "api.moby.localhost"
	targetProtocol  = "http://"
)

func MyConcat(x string) string {
    w := targetProtocol + hostHeaderValue +  x
    return w
}
```

it is basically compiled to Go Assembler as this:

```
        LEAQ    go:string."http://api.moby.localhost"(SB), BX
        PCDATA  $1, $1
        CALL    runtime.concatstring2(SB)
```

Nino-K reviewed (commented):
@bcxpro I'm just going to do a little more testing and then will approve and merge this.
